README.md
# DEPRECATED

**This version of Lute is obsolete and will no longer be maintained.  Instead, please use [Lute v3](https://github.com/jzohrab/lute-v3)**

Lute v3 is a full Python rewrite of Lute.  Installation is _much_ easier, and all the features have been ported.

- [Install Lute v3](https://jzohrab.github.io/lute-manual/install.html)
- [lute-v3 repository](https://github.com/jzohrab/lute-v3) and [manual](https://jzohrab.github.io/lute-manual/intro.html)
- If you have Lute v2, you can easily [migrate to v3](https://jzohrab.github.io/lute-manual/v2-to-v3.html).
  
_Thank you!_

---

# LUTE - Learning Using Texts

[![CI master](https://github.com/jzohrab/lute/actions/workflows/symfony-ci.yml/badge.svg?branch=master)](https://github.com/jzohrab/lute/actions/workflows/symfony-ci.yml?query=branch%3Amaster)
[![Discord Server](https://badgen.net/badge/icon/discord?icon=discord&label)](https://discord.gg/CzFUQP5m8u)
[![No Maintenance Intended](http://unmaintained.tech/badge.svg)](http://unmaintained.tech/)

This is a fork and ground-up rewrite of Learning with Texts, for learning foreign languages through reading.  You install Lute on your computer, import texts, and annotate terms.

Lute contains the core features you need for learning through reading:

* defining languages and dictionaries
* creating and editing texts
* creating terms and multi-word terms

In addition, Lute has features missing from other tools, which I consider absolutely essential:

* Defining "parent terms".  For example, for Spanish, _hablar_ (to speak) is the root form of _habló_ (he spoke).  In Lute, you can set a root form as the Parent Term, and the definition for that root form is available for all child terms.  Also, when you ask Lute for sample sentences using a term, it includes sentences for all related terms as well.  See [the wiki](https://github.com/jzohrab/lute/wiki/Parent-Terms) for examples.
* Adding word images.  This helps learning, and is more interesting and entertaining than adding reams of text.

I have been using Lute for learning Spanish, it's a super tool.

## Demo

A _very_ brief demo, showing the core feature: reading a doc, and creating a term:

[comment]: # (See docs/adding_readme_gif.md for notes)

![A wee demo](https://user-images.githubusercontent.com/1637133/210660839-b9aebebc-60c6-43fc-9f6d-daf2c448f825.gif)

_(fyi - The screenshot was edited for time and file size, so it looks like Lute is automatically filling in the term form -- it's not.)_

Lute has other features not shown. See the [YouTube introduction video](https://youtu.be/cjSqQTwUFCY) -- it's a bit out-of-date, but gives you an idea -- and check out the [Wiki](https://github.com/jzohrab/lute/wiki).

# Installation

The easiest way to get started with Lute is using Docker.  See [Running Lute using Docker](https://github.com/jzohrab/lute/wiki/Getting-Started) in the Lute Wiki.

If you can't use Docker, you can also [install everything manually](https://github.com/jzohrab/lute/wiki/Getting-started-without-Docker), which is how I run Lute myself on my Mac.

# Docs

Docs are in the [Wiki](https://github.com/jzohrab/lute/wiki).

You can also join [the Lute Discord Server](https://discord.gg/CzFUQP5m8u).

# Contribution

If you're a gearhead, like me:

* Read the [development notes](https://github.com/jzohrab/lute/wiki/Development)
* Fork this repo
* Run `composer install --dev` to install dependencies.  If running acceptance tests, run `vendor/bin/bdi detect drivers` to install Panther drivers
* Make your changes
* Run `composer test:full` to ensure everything passes.  This runs psalm, Doctrine ORM mapping checks, and unit and acceptance tests
* Open a PR to the `develop` branch

If you're a user: Lute is free :-) ... but if it makes your life better and you feel like saying thanks, I gladly accept <a href="https://www.buymeacoffee.com/jzohrab" target="_blank">coffee</a>.  I'll give thanks and will use the caffeine to implement features, or, better, recruit devs to grow the project.

# Unlicense

Under unlicense, view [UNLICENSE.md](UNLICENSE.md), and check out [http://unlicense.org/].

# Acknowledgements

Lute would never have existed without the original [Learning With Texts](https://sourceforge.net/projects/learning-with-texts) and [Hugo Fara's fork](https://github.com/hugofara), so **a big thanks to both of these projects.**


docs/CHANGELOG.md
# Changelog

## 2023-11-11 v2.1.4_EndOfLife

**This is the END-OF-LIFE release for Lute v2.**

Lute v3 is out, and Lute v2 will no longer be maintained.

If this is your first time installing Lute, then
you should instead install **Lute v3**, see the docs at:

https://jzohrab.github.io/lute-manual/install.html

_This release is merely done to ship the rest of the
v2 code, for users who won't update to v3 for
some reason._

Feature changes:

* None.

Back-end changes:

There are _many_ back-end changes that facilitated the port to Python:

* Change reading URL to /read/<book_id>/page/<page_number>
* Port all tests
* Remove unused code
* Simplify some schema
* Add smoke tests for coverage.
* Query env. via api to see if mecab installed.
* Use primitives not entities in web tests.
* Extract common code.
* Save sentences when saving Text.
* Fix namespaces.


## 2023-10-07 v2.1.3

Features:

* In reading screen Term form, clicking on parent tag now moves to the parent word.  For example, if I have a term "dogs" with parent "dog", clicking on the "dog" parent tag will display the form for "dog".  Ref the [wiki](https://github.com/jzohrab/lute/wiki/Parent-Terms#click-a-parent-tag-to-edit-the-parent-term)
* Images now saved and searched by parent text.  For example, if the Term is "dogs", and the parent is "dog", "dog" will be searched.

Fixes:

* Issue [77](https://github.com/jzohrab/lute/issues/77): Turkish term case changes.  Old code was using naive javascript, now posts back to server.
* Issue [75](https://github.com/jzohrab/lute/issues/75): Stop images moving around on image search

Back end changes:

* Some changes to structure to clarify the migration to Lute v3: add Parse and Render namespaces, remove unused code.


## 2023-10-05 v2.1.2

Feature changes:

* Added support for Turkish (issue [71](https://github.com/jzohrab/lute/issues/71)).  There's a new "parse as" "turkish" Language option, because Turkish has a few characters for "i" (i, İ; ı, I) that need special handling.

Bug fixes:

* Prevent frozen Term hover (issue [71](https://github.com/jzohrab/lute/issues/73))

Back end changes:

* Reorganize demo data and language templates in demo/
* Load language from yaml
* Fix Windows GitHub CI tests


## 2023-09-29 v2.1.1

Feature changes:

* Add Right-to-Left language support and a sample Arabic text (issue [67](https://github.com/jzohrab/lute/issues/67))

Back end:

* Re-activate acceptance tests for develop, master


## 2023-09-22 v2.1.0

Various fixes for new user non-Docker installs:

* Windows users only: Add .env file variable `DB_WINDOWS_REPLACE_TRIPLE_SLASH=yes`.  If you're running Windows and get the error "SQLSTATE[HY000] [14] unable to open database file", try using this .env setting.
* Remove Japanese from new db demo data if user doesn't have MeCab (issue [69](https://github.com/jzohrab/lute/issues/69))


## 2023-09-17 v2.0.14

This release preserves the case of terms created from texts.  For example, clicking "Paris" in "Je suis à Paris" now creates the term "Paris", instead of "paris".  You can edit the term to change the letter capitalization if needed (but you can't change the term itself -- see [the wiki](https://github.com/jzohrab/lute/wiki/Terms-cannot-be-changed-once-created) for notes).

Feature changes:

* Preserve case for terms when creating from texts (issue [63](https://github.com/jzohrab/lute/issues/63))
* Allow changing case of Term letters 
* Allow spaces in parent tags (hit Return after entering a parent)
* Fix text moving/shifting on hover/click

## 2023-08-28 v2.0.13

This release adds support for multiple parents, and simplifies some back-end data storage.

Feature changes:

* Allow terms to have multiple parents ([issue 61](https://github.com/jzohrab/lute/issues/61))
* Enter parents using tags
* Term detail pop-up (on reading page) modified to display data for multiple parents
* Sentence lookups return sentences for multiple parents

Back end changes:

* Remove `texttokens` table - big space savings
* Change popup to API endpoint
* Simplify reading page rendering, use fewer attributes
* Store word count in book table to speed stats calc


## 2023-08-17 v2.0.12

Feature changes:

* Remove spaces between rendered spans to support [Zhongwen Chrome extension](https://github.com/cschiller/zhongwen)
* Fix half-covered "?" help icon when term form opens

Back end changes:

* remove unused code, simplify existing.


## 2023-08-04 v2.0.11

Feature changes:

* Fix missing -ru character at end of Japanese terms ([issue 60](https://github.com/jzohrab/lute/issues/60)).

## 2023-08-03 v2.0.10

Feature changes:

* Fix stats for character-based languages ([issue 55](https://github.com/jzohrab/lute/issues/55)), with notes on [wiki](https://github.com/jzohrab/lute/wiki/Stats-calculation)
* Fix Japanese paragraph copy ([issue 56](https://github.com/jzohrab/lute/issues/56)), adding paragraph numbers to rendering
* Remove "hasflash" style on term click

Back end changes:

* Speed up RenderableCalculator, TokenLocator


## 2023-07-24 v2.0.9

Feature changes:

* Hotkey shift + t = translate full paragraph.
* Don't copy to clipboard when making multiterm.
* Add book reparsing (lightning bolt icon) - use this only when changing your Language parse settings
* Fix Japanese sentence end finding.

Back end changes:

* Simplifying and removing code where possible!


## 2023-07-15 v2.0.8

Feature changes:

* Change hotkeys to also work on currently hovered word (issue [46](https://github.com/jzohrab/lute/issues/46))
* Add "shift-drag" copy-to-clipboard
* Update tutorial to explain new hotkey handling (included in [the wiki](https://github.com/jzohrab/lute/wiki/Tutorial))

Back end changes:

* Add hacky lute.js cachebusting, so that browsers serve correct Javascript on Lute update!
* Hotkey acceptance tests


## 2023-07-10 v2.0.7

Feature changes:


public/js/lute.js
/* Lute functions. */

/**
 * The current term index (either clicked or hovered over)
 */
let LUTE_CURR_TERM_DATA_ORDER = -1;  // initially not set.

/**
 * Lute has 2 different "modes" when reading:
 * - LUTE_HOVERING = true: Hover mode, not selecting
 * - LUTE_HOVERING = false: Word clicked, or click-drag
 */ 
let LUTE_HOVERING = true;


/**
 * When the reading pane is first loaded, it is set in "hover mode",
 * meaning that when the user hovers over a word, that word becomes
 * the "active word" -- i.e., status update keyboard shortcuts should
 * operate on that hovered word, and as the user moves the mouse
 * around, the "active word" changes.  When a word is clicked, though,
 * there can't be any "hover changes", because the user should be
 * editing the word in the Term edit pane, and has to consciously
 * disable the "clicked word" mode by hitting ESC or RETURN.
 *
 * When the user is done editing a the Term form in the Term edit pane
 * and hits "save", the main reading page's text div is updated (see
 * templates/read/updated.twig.html).  This text div reload then has
 * to notify _this_ javascript to start_hover_mode() again.
 * 
 * I dislike this code (specifically, that the updated.twig.html calls
 * this javascript function), but can't think of a better way to
 * manage this.
 */
function start_hover_mode(should_clear_frames = true) {
  // console.log('CALLING RESET');
  load_reading_pane_globals();
  LUTE_HOVERING = true;

  $('span.kwordmarked').removeClass('kwordmarked');

  const w = get_current_word();
  if (w != null) {
    $(w).addClass('wordhover');
  }

  if (should_clear_frames)
    clear_frames();

  clear_newmultiterm_elements();

  // https://stackoverflow.com/questions/35022716/keydown-not-detected-until-window-is-clicked
  $(window).focus();
}

let clear_frames = function() {
  $('#wordframeid').attr('src', '/read/empty');
  $('#dictframeid').attr('src', '/read/empty');
}

/** 
 * Prepare the interaction events with the text.
 */
function prepareTextInteractions(textid) {
  const t = $('#thetext');
  // Using "t.on" here because .word elements
  // are added and removed dynamically, and "t.on"
  // ensures that events remain for each element.
  t.on('mousedown', '.word', select_started);
  t.on('mouseover', '.word', select_over);
  t.on('mouseup', '.word', select_ended);

  t.on('mouseover', '.word', hover_over);
  t.on('mouseout', '.word', hover_out);
  
  $(document).on('keydown', handle_keydown);

  $('#thetext').tooltip({
    position: { my: 'left top+10', at: 'left bottom', collision: 'flipfit flip' },
    items: '.word.showtooltip',
    show: { easing: 'easeOutCirc' },
    content: function () { return tooltip_textitem_hover_content($(this)); }
  });
}


/**
 * Build the html content for jquery-ui tooltip.
 */
let tooltip_textitem_hover_content = function (el) {
  elid = parseInt(el.attr('data_wid'));
  let content = null;
  $.ajax({
    url: `/termpopup/${elid}`,
    type: 'get',
    async: false,
    success: function(response) {
      content = response;
    }
  });
  return content;
}


function showEditFrame(el, extra_args = {}) {
  const lid = parseInt(el.attr('lid'));

  let text = extra_args.textparts ?? [ el.attr('data_text') ];
  const sendtext = text.join('');

  let extras = Object.entries(extra_args).
      map((p) => `${p[0]}=${encodeURIComponent(p[1])}`).
      join('&');

  // Annoying hack.  I wanted to send a period '.' in the term, but
  // Symfony didn't handle that well (per issue
  // https://github.com/jzohrab/lute/issues/28), and replacing the '.'
  // with the encoded value ('\u2e' or '%2E') didn't work, as it kept
  // getting changed back to '.' on send, or it said that it couldn't
  // find the controller route.  There is probably something very
  // simple I'm missing, but for now, replacing the '.' with a hacky
  // string which I'll replace on server side as well.
  const periodhack = '__LUTE_PERIOD__';
  const url = `/read/termform/${lid}/${sendtext}?${extras}`.replaceAll('.', periodhack);
  // console.log('go to url = ' + url);

  top.frames.wordframe.location.href = url;
}


function show_help() {
  const url = `/read/shortcuts`;
  top.frames.wordframe.location.href = url;
}


let save_curr_data_order = function(el) {
  LUTE_CURR_TERM_DATA_ORDER = parseInt(el.attr('data_order'));
}

/* ========================================= */
/** Hovering */

function hover_over(e) {
  if (! LUTE_HOVERING)
    return;
  $('span.wordhover').removeClass('wordhover');
  $(this).addClass('wordhover');
  save_curr_data_order($(this));
}

function hover_out(e) {
  if (! LUTE_HOVERING)
    return;
  $('span.wordhover').removeClass('wordhover');
}

/* ========================================= */
/** Multiword selection */

let selection_start_el = null;

let clear_newmultiterm_elements = function() {
  $('.newmultiterm').removeClass('newmultiterm');
  selection_start_el = null;
}

function select_started(e) {
  LUTE_HOVERING = false;
  $('span.wordhover').removeClass('wordhover');
  clear_newmultiterm_elements();
  clear_frames();
  $(this).addClass('newmultiterm');
  selection_start_el = $(this);
  save_curr_data_order($(this));
}

let get_selected_in_range = function(start_el, end_el, selector) {
  const first = parseInt(start_el.attr('data_order'))
  const last = parseInt(end_el.attr('data_order'));

  let startord = first;
  let endord = last;

  if (startord > endord) {
    endord = first;
    startord = last;
  }

  const selected = $(selector).filter(function() {
    const ord = $(this).attr("data_order");
    return ord >= startord && ord <= endord;
  });
  return selected;
};

function select_over(e) {
  if (selection_start_el == null)
    return;  // Not selecting
  $('.newmultiterm').removeClass('newmultiterm');
  const selected = get_selected_in_range(selection_start_el, $(this), 'span.textitem');
  selected.addClass('newmultiterm');
}

function select_ended(e) {
  // Handle single word click.
  if (selection_start_el.attr('id') == $(this).attr('id')) {
    clear_newmultiterm_elements();
    word_clicked($(this), e);
    return;
  }

  $('span.kwordmarked').removeClass('kwordmarked');

  const selected = get_selected_in_range(selection_start_el, $(this), 'span.textitem');
  if (e.shiftKey) {
    copy_text_to_clipboard(selected.toArray());
    start_hover_mode(false);
    return;
  }

  const textparts = selected.toArray().map((el) => $(el).text());
  const text = textparts.join('').trim();
  if (text.length > 250) {
    alert(`Selections can be max length 250 chars ("${text}" is ${text.length} chars)`);
    start_hover_mode();
    return;
  }

  showEditFrame(selection_start_el, { textparts: textparts });
  selection_start_el = null;
}


let word_clicked = function(el, e) {
  el.removeClass('wordhover');
  save_curr_data_order(el);
  if (el.hasClass('kwordmarked')) {
    el.removeClass('kwordmarked');
    const nothing_marked = $('span.kwordmarked').length == 0;
    if (nothing_marked) {
      el.addClass('wordhover');
      start_hover_mode();
    }
  }
  else {
    if (! e.shiftKey) {
      $('span.kwordmarked').removeClass('kwordmarked');
      showEditFrame(el);
    }
    el.addClass('kwordmarked');
    el.removeClass('hasflash');
  }
}


/********************************************/
// Keyboard navigation.

// Load all words into scope.
var words = null;
var maxindex = null;

function load_reading_pane_globals() {
  words = $('span.word').sort(function(a, b) {
    return $(a).attr('data_order') - $(b).attr('data_order');
  });
  // console.log('have ' + words.size() + ' words');
  maxindex = words.size() - 1;
}

$(document).ready(load_reading_pane_globals);

let current_word_index = function() {
  const i = words.toArray().findIndex(x => parseInt(x.getAttribute('data_order')) === LUTE_CURR_TERM_DATA_ORDER);
  // console.log(`found index = ${i}`);
  return i;
};

let get_current_word = function() {
  const selindex = current_word_index();
  if (selindex == -1)
    return null;
  return words.eq(selindex);
}

/** Get the rest of the textitems in the current active/hovered word's
 * sentence or paragraph, or null if no selection. */
let get_textitems_spans = function(e) {
  const w = get_current_word();
  if (w == null)
    return null;

  let attr_name = 'seid';
  let attr_value = w.attr('seid');

  if (e.shiftKey) {
    attr_name = 'paraid';
    attr_value = w.attr('paraid');
  }

  return $('span.textitem').toArray().filter(x => x.getAttribute(attr_name) === attr_value);
};

/** Copy the text of the textitemspans to the clipboard, and add a
 * color flash. */
let handle_copy = function(e) {
  tis = get_textitems_spans(e);
  if (tis == null)
    return;
  copy_text_to_clipboard(tis);
}

let copy_text_to_clipboard = function(textitemspans, show_flash = true) {
  const copytext = textitemspans.map(s => $(s).text()).join('');

  // console.log('copying ' + copytext);
  var textArea = document.createElement("textarea");
  textArea.value = copytext;
  document.body.appendChild(textArea);
  textArea.select();
  document.execCommand("Copy");
  textArea.remove();

  const removeFlash = function() {
    // console.log('removing flash');
    $('span.flashtextcopy').addClass('wascopied'); // for acceptance testing.
    $('span.flashtextcopy').removeClass('flashtextcopy');
  };

  // Add flash, set timer to remove.
  removeFlash();
  textitemspans.forEach(function (t) {
    $(t).addClass('flashtextcopy');
  });
  setTimeout(() => removeFlash(), 1000);

  $('#wordframeid').attr('src', '/read/flashcopied');
}


let set_cursor = function(newindex) {
  LUTE_HOVERING = false;
  $('span.wordhover').removeClass('wordhover');
  clear_newmultiterm_elements();

  if (newindex < 0 || newindex >= words.size())
    return;
  let curr = words.eq(newindex);
  save_curr_data_order(curr);
  $('span.kwordmarked').removeClass('kwordmarked');
  curr.addClass('kwordmarked');
  $(window).scrollTo(curr, { axis: 'y', offset: -150 });
  showEditFrame(curr, { autofocus: false });
}


let find_non_Ign_or_Wkn = function(currindex, shiftby) {
  let newindex = currindex + shiftby;
  while (newindex >= 0 && newindex <= maxindex) {
    const nextword = words.eq(newindex);
    const st = nextword.attr('data_status');
    if (st != 99 && st != 98) {
      break;
    }
    newindex += shiftby;
  }
  return newindex;
};

let move_cursor = function(shiftby, e) {
  const currindex = current_word_index();
  if (! e.shiftKey) {
    set_cursor(currindex + shiftby);
  }
  else {
    set_cursor(find_non_Ign_or_Wkn(currindex, shiftby));
  }
}


let show_translation = function(e) {
  tis = get_textitems_spans(e);
  if (tis == null)
    return;
  const sentence = tis.map(s => $(s).text()).join('');
  // console.log(sentence);

  const userdict = $('#translateURL').text();
  if (userdict == null || userdict == '')
    console.log('No userdict for lookup.  ???');

  // console.log(userdict);
  const url = userdict.replace('###', encodeURIComponent(sentence));
  if (url[0] == '*') {
    const finalurl = url.substring(1);  // drop first char.
    const settings = 'width=800, height=400, scrollbars=yes, menubar=no, resizable=yes, status=no';
    window.open(finalurl, 'dictwin', settings);
  }
  else {
    top.frames.dictframe.location.href = url;
  }
}


function handle_keydown (e) {
  if (words.size() == 0) {
    // console.log('no words, exiting');
    return; // Nothing to do.
  }

  // Map of key codes (e.which) to lambdas:
  let map = {};

  const kESC = 27;
  const kRETURN = 13;
  const kHOME = 36;
  const kEND = 35;
  const kLEFT = 37;
  const kRIGHT = 39;
  const kC = 67; // C)opy
  const kT = 84; // T)ranslate
  const k1 = 49;
  const k2 = 50;
  const k3 = 51;
  const k4 = 52;
  const k5 = 53;
  const kI = 73;
  const kW = 87;

  map[kESC] = () => start_hover_mode();
  map[kRETURN] = () => start_hover_mode();
  map[kHOME] = () => set_cursor(0);
  map[kEND] = () => set_cursor(maxindex);
  map[kLEFT] = () => move_cursor(-1, e);;
  map[kRIGHT] = () => move_cursor(+1, e);;
  map[kC] = () => handle_copy(e);
  map[kT] = () => show_translation(e);
  map[k1] = () => update_status_for_marked_elements(1);
  map[k2] = () => update_status_for_marked_elements(2);
  map[k3] = () => update_status_for_marked_elements(3);
  map[k4] = () => update_status_for_marked_elements(4);
  map[k5] = () => update_status_for_marked_elements(5);
  map[kI] = () => update_status_for_marked_elements(98);
  map[kW] = () => update_status_for_marked_elements(99);

  if (e.which in map) {
    let a = map[e.which];
    a();
  }
  else {
    // console.log('unhandled key ' + e.which);
  }
}


/**
 * post update ajax call, fix the UI.
 */
function update_selected_statuses(newStatus) {
  const newClass = `status${newStatus}`;
  let update_status = function (e) {
    const curr = $(this);
    ltext = curr.text().toLowerCase();
    matches = $('span.word').toArray().filter(el => $(el).text().toLowerCase() == ltext);
    matches.forEach(function (m) {
      $(m).removeClass('status98 status99 status0 status1 status2 status3 status4 status5 shiftClicked')
        .addClass(newClass)
        .attr('data_status',`${newStatus}`);
    });
  };
  $('span.kwordmarked').each(update_status);
  $('span.wordhover').each(update_status);
}


function update_status_for_marked_elements(new_status) {
  let els = $('span.kwordmarked').toArray().concat($('span.wordhover').toArray());
  if (els.length == 0)
    return;
  const firstel = $(els[0]);
  const textid = firstel.attr('tid');

  els = els.map(el => $(el).text());

  $.ajax({
    url: '/read/update_status',
    type: 'post',
    data: { textid: textid, terms: els, new_status: new_status },
    dataType: 'JSON',
    success: function(response) {
      update_selected_statuses(new_status);
    },
    error: function(response, status, err) {
      const msg = {
        response: response,
        status: status,
        error: err
      };
      console.log(`failed: ${JSON.stringify(msg, null, 2)}`);
    }
  });

}
